cw/cw305/simple_capture_traces.py

```diff
@@ -40,101 +55,153 @@ def initialize_capture(device_cfg, capture_cfg):
     return ot
 
 
-def run_capture(capture_cfg, ot, ktp):
-    """Run ChipWhisperer capture.
+def plot_results(plot_cfg, project_name):
+  """Plots traces from `project_name` using `plot_cfg` settings."""
+  project = cw.open_project(project_name)
 
-    Args:
-      capture_cfg: Dictionary with capture configuration settings.
-      ot: Initialized OpenTitan target.
-      ktp: Key and plaintext generator.
-    """
-    key, text = ktp.next()
+  if len(project.waves) == 0:
+    print('Project contains no traces. Did the capture fail?')
+    return
 
-    cipher = AES.new(bytes(key), AES.MODE_ECB)
-    print(f'Using key: {binascii.b2a_hex(bytes(key))}')
+  # The ADC output is in the interval [-0.5, 0.5). Check that the recorded
+  # traces are within that range with some safety margin.
+  if not (np.all(np.greater(project.waves, -plot_cfg.amplitude_max)) and
+          np.all(np.less(project.waves, plot_cfg.amplitude_max))):
+    print('WARNING: Some traces have samples outside the range (' +
+          str(-plot_cfg.amplitude_max) + ', ' +
+          str(plot_cfg.amplitude_max) + ').')
+    print('         The ADC has a max range of [-0.5, 0.5) and might saturate.')
+    print('         It is recommended to reduce the scope gain (see device.py).')
 
-    project_file = capture_cfg['project_name']
-    project = cw.create_project(project_file, overwrite=True)
+  plot.save_plot_to_file(project.waves, plot_cfg.num_traces,
+                         plot_cfg.trace_image_filename)
+  print(f'Created plot: {Path(plot_cfg.trace_image_filename).resolve()}')
 
-    for i in tqdm(range(capture_cfg['num_traces']), desc='Capturing', ncols=80):
-        key, text = ktp.next()
-        ret = cw.capture_trace(ot.scope, ot.target, text, key, ack=False)
-        if not ret:
-            print('Failed capture')
-            continue
 
-        expected = binascii.b2a_hex(cipher.encrypt(bytes(text)))
-        got = binascii.b2a_hex(ret.textout)
-        assert (got == expected), (
-            f'Incorrect encryption result!\ngot: {got}\nexpected: {expected}\n'
-        )
+@app.command()
+def init(ctx: typer.Context):
+    """Initalize target for SCA."""
+    initialize_capture(ctx.obj.cfg.device, ctx.obj.cfg.capture)
+
+
+def capture_init(ctx, num_traces, plot_traces):
+    """Initializes the user data stored in the context and programs the target."""
+    cfg = ctx.obj.cfg
+    if num_traces:
+        cfg.capture.num_traces = num_traces
+
+    if plot_traces:
+        cfg.plot_capture.show = True
+        cfg.plot_capture.num_traces = plot_traces
 
-        project.traces.append(ret)
+    # Key and plaintext generator
+    ctx.obj.ktp = cw.ktp.Basic()
+    ctx.obj.ktp.key_len = cfg.capture.key_len_bytes
+    ctx.obj.ktp.text_len = cfg.capture.plain_text_len_bytes
+
+    ctx.obj.ot = initialize_capture(cfg.device, cfg.capture)
 
-        ot.target.flush()
 
+def capture_loop(traces, capture_cfg):
```

Comment:
If I understand this correctly, `traces` is actually a function including arguments. Inside the loop, you then just call that over and over again. I like this as it allows for sharing the logic around for mutliple capture functions.

But what I don't understand:
- What is `next()` doing on a function? Is it just running it once more, or is there some more magic going on? 
- Inside the `traces` function, you are updating `ktp`. Does that mean every trace is collected with a different key?

Suggestions:
- I would give `traces` a different name to indicate that it's actually a function capturing a trace.
- Maybe add another, optional function to initialize the reference cipher such that it doesn't have to be done inside the capture function upon every call.

Reply:
These explanations are very helpful, thanks @alphan! (also for renaming `traces` to `trace_gen`).